pkg/action/upgrade-recovery.go

```diff
@@ -146,48 +146,73 @@ func (u *UpgradeRecoveryAction) Run() (err error) {
 		return err
 	}
 
-	// Create recovery /boot dir if not exists
-	bootDir := filepath.Join(u.spec.Partitions.Recovery.MountPoint, "boot")
-	if err := utils.MkdirAll(u.cfg.Fs, bootDir, constants.DirPerm); err != nil {
-		u.cfg.Logger.Errorf("failed creating recovery boot dir: %v", err)
-		return elementalError.NewFromError(err, elementalError.CreateDir)
+	// Remove any traces of previously errored upgrades
+	transitionDir := filepath.Join(u.spec.Partitions.Recovery.MountPoint, constants.BootTransitionDir)
+	if ok, _ := utils.Exists(u.cfg.Fs, transitionDir); ok {
+		u.Debugf("removing orphaned recovery system %s", transitionDir)
+		err = u.cfg.Fs.RemoveAll(transitionDir)
+		if err != nil {
+			u.Errorf("failed removing old recovery image: %s", err.Error())
+			return err
+		}
 	}
 
-	// Upgrade recovery
-	err = elemental.DeployRecoverySystem(u.cfg.Config, &u.spec.RecoverySystem, bootDir)
+	// Deploy recovery system to transition dir
+	err = elemental.DeployRecoverySystem(u.cfg.Config, &u.spec.RecoverySystem)
 	if err != nil {
-		u.cfg.Logger.Errorf("failed deploying recovery image: %v", err)
+		u.cfg.Logger.Errorf("failed deploying recovery image: %s", err.Error())
 		return elementalError.NewFromError(err, elementalError.DeployImage)
 	}
-	recoveryFile := filepath.Join(u.spec.Partitions.Recovery.MountPoint, constants.RecoveryImgFile)
-	transitionFile := filepath.Join(u.spec.Partitions.Recovery.MountPoint, constants.TransitionImgFile)
-	if ok, _ := utils.Exists(u.cfg.Fs, recoveryFile); ok {
-		err = u.cfg.Fs.Remove(recoveryFile)
+
+	// Switch places on /boot and transition-dir
+	existingDir := filepath.Join(u.spec.Partitions.Recovery.MountPoint, constants.BootDir)
+	oldBootDir := filepath.Join(u.spec.Partitions.Recovery.MountPoint, constants.OldBootDir)
+	if ok, _ := utils.Exists(u.cfg.Fs, existingDir); ok {
+		err = u.cfg.Fs.Rename(existingDir, oldBootDir)
 		if err != nil {
-			u.Error("failed removing old recovery image")
+			u.Errorf("failed removing old recovery image: %s", err.Error())
 			return err
 		}
 	}
-	err = u.cfg.Fs.Rename(transitionFile, recoveryFile)
+	err = u.cfg.Fs.Rename(transitionDir, existingDir)
 	if err != nil {
-		u.Error("failed renaming transition recovery image")
+		u.cfg.Logger.Errorf("failed renaming transition recovery image: %s", err.Error())
+
+		// Try to salvage old recovery system
+		if ok, _ := utils.Exists(u.cfg.Fs, oldBootDir); ok {
+			err = u.cfg.Fs.Rename(existingDir, oldBootDir)
+			if err != nil {
+				u.cfg.Logger.Errorf("failed salvaging old recovery system: %s", err.Error())
+			}
+		}
+
 		return err
 	}
 
+	// Remove old boot-dir when new recovery system is in place
+	if ok, _ := utils.Exists(u.cfg.Fs, oldBootDir); ok {
+		err = u.cfg.Fs.RemoveAll(oldBootDir)
+		if err != nil {
+			u.Errorf("failed removing old recovery image: %s", err.Error())
+			return err
```

Comment:
I am wondering if we should actually return on this error. If we return here this ends up in a very weird state, as it reports error and abruptly finishes but the upgrade is indeed already done 🤔 On reboot it will be a there actually.

Probably once the swap is successfully done we should not error out on errors to give chance to continue with any remaining step. 

Anyway this might be a separate issue as this would also apply for regular upgrades. At that point we could eventually collect relevant errors but continue with the execution until the end.

Reply:
It's a good point, we should also remove the oldBootDir on newer upgrades if it already exists!